### PR TITLE
[REVIEW] Refactoring cooperative loading with single thread loading

### PR DESCRIPTION
**Pull request by rgsl888prabhu**
I have tried trove, but there wasn't much of a gain compared to single threaded usage and didn't loose any performance compared to previous version of CUIO.
<details>
<summary>Perf plots</summary>

y-axis is in `ms` and there are two sets of plot, one which compares performance change next to each other and other one which difference of performance between single thread and coalesced loading.  If the value is positive, then test is taking more time, else it is taking lesser time compared to coalesced loading.

ORC_READER_0_32
ORC_READER_0_32_CH
ORC_READER_32_64
ORC_READER_32_64_CH
ORC_WRITER_0_32
ORC_WRITER_0_32_CH
ORC_WRITER_32_64
ORC_WRITER_32_64_CH
ORC_WRITER_64_93
ORC_WRITER_64_93_CH
PARQUET_CHUNK_WRITER
PARQUET_CHUNK_WRITER_CH
PARQUET_READER_0_32
PARQUET_READER_0_32_CH
PARQUET_READER_32_64
PARQUET_READER_32_64_CH
PARQUET_READER_64_80
PARQUET_READER_64_80_CH
PARQUET_WRITER_0_32
PARQUET_WRITER_0_32_CH
PARQUET_WRITER_32_64
PARQUET_WRITER_32_64_CH
PARQUET_WRITER_64_96
PARQUET_WRITER_64_96_CH
PARQUET_WRITER_96_120
PARQUET_WRITER_96_120_CH

</details>

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: cpp/src/io/parquet/page_data.cu [hunk 1]
```diff
@@ -912,16 +912,14 @@ static __device__ bool setupLocalPageInfo(page_state_s *const s,
...
-  if (t < sizeof(PageInfo) / sizeof(uint32_t)) {
-    reinterpret_cast<uint32_t *>(&s->page)[t] = reinterpret_cast<const uint32_t *>(p)[t];
-  }
+  if (t == 0) s->page = *p;
   __syncthreads();
+
   if (s->page.flags & PAGEINFO_FLAGS_DICTIONARY) { return false; }
   // Fetch column chunk info
   chunk_idx = s->page.chunk_idx;
+  // TO DO, need to replace, currently has some race condition
```
Comment (devavret) [inline cpp/src/io/parquet/page_data.cu:921]: ideally every such load should be succeeded by a syncthreads

**Review by devavret — CHANGES_REQUESTED** [review]
a simple syncthreads should fix this

File: cpp/src/io/parquet/page_data.cu (hunk 1 above)
Comment (rgsl888prabhu) [reply cpp/src/io/parquet/page_data.cu:921]: There is a syncthreads below at line 938, and as this parameter is not being used in the middle thought it wouldn't be of any use to call one more syncthreads.

File: cpp/src/io/orc/dict_enc.cu [hunk 2]
```diff
@@ -127,10 +127,8 @@ __global__ void __launch_bounds__(block_size, 2)
   uint32_t nnz, start_row, dict_char_count;
   int t = threadIdx.x;
 
-  if (t < sizeof(DictionaryChunk) / sizeof(uint32_t)) {
-    ((volatile uint32_t *)&s->chunk)[t] =
-      ((const uint32_t *)&chunks[group_id * num_columns + col_id])[t];
-  }
+  if (t == 0) s->chunk = chunks[group_id * num_columns + col_id];
+  // TO-DO, if possible have to simplify
```
Comment (devavret) [inline cpp/src/io/orc/dict_enc.cu:131]: Not really. This is a hash map being initialized. It's different from cooperative load/store and is scope creep. It will probably be refactored out in a separate PR when we abstract away hash maps.

File: cpp/src/io/parquet/page_data.cu [hunk 3]
```diff
@@ -1127,7 +1119,7 @@ static __device__ bool setupLocalPageInfo(page_state_s *const s,
         s->page.skipped_leaf_values = s->first_row;
       }
 
-      s->input_value_count = s->page.skipped_values;
+      s->input_value_count = s->page.skipped_values > -1 ? s->page.skipped_values : 0;
```
Comment (rgsl888prabhu) [inline cpp/src/io/parquet/page_data.cu:1122]: This was causing race condition where default `s->page.skipped_values` is `-1` which gets assigned to `s->input_value_count`  which is supposed to be `0`. Then while indexing with thread 0, -1 will be converted to 255, causing access of improper value [here](https://github.com/rapidsai/cudf/blob/d1e8fee20b1c25059d5d3a35e7bd862fa0d6c6d9/cpp/src/io/parquet/page_data.cu#L1225)

Comment (rgsl888prabhu) [reply cpp/src/io/parquet/page_data.cu:1122]: https://github.com/rapidsai/cudf/blob/d1e8fee20b1c25059d5d3a35e7bd862fa0d6c6d9/cpp/src/io/parquet/page_data.cu#L1225

File: cpp/src/io/orc/stripe_init.cu [hunk 4]
```diff
@@ -32,14 +32,12 @@ extern "C" __global__ void __launch_bounds__(128, 8) gpuParseCompressedStripeDat
 {
   __shared__ compressed_stream_s strm_g[4];
 
-  volatile compressed_stream_s *const s = &strm_g[threadIdx.x >> 5];
-  int strm_id                           = blockIdx.x * 4 + (threadIdx.x >> 5);
-  int t                                 = threadIdx.x & 0x1f;
+  compressed_stream_s *const s = &strm_g[threadIdx.x >> 5];
+  int strm_id                  = blockIdx.x * 4 + (threadIdx.x >> 5);
+  int t                        = threadIdx.x & 0x1f;
```
Comment (devavret) [inline cpp/src/io/orc/stripe_init.cu:37]: This means `t` is lane id. Is it?

Comment (rgsl888prabhu) [reply cpp/src/io/orc/stripe_init.cu:37]: Yes, it is lane id. I will update it with appropriate name.

File: cpp/src/io/orc/stripe_init.cu [hunk 5]
```diff
@@ -127,14 +119,12 @@ extern "C" __global__ void __launch_bounds__(128, 8)
...
```
Comment (devavret) [inline cpp/src/io/orc/stripe_init.cu:124]: Same here. What's going on? Was this a latent bug

File: cpp/src/io/parquet/page_enc.cu [hunk 6]
```diff
@@ -1444,20 +1404,15 @@ __global__ void __launch_bounds__(128) gpuEncodePageHeaders(EncPage *pages,
...
-      reinterpret_cast<const uint32_t *>(&chunks[page_g.chunk_id])[t];
-  }
+
+  if (t == 0) ck_g = chunks[page_g.chunk_id];
   __syncthreads();
-  if (t < sizeof(EncColumnDesc) / sizeof(uint32_t)) {
-    reinterpret_cast<uint32_t *>(&col_g)[t] = reinterpret_cast<const uint32_t *>(ck_g.col_desc)[t];
-  }
+
+  if (t == 0) col_g = *ck_g.col_desc;
```
Comment (devavret) [inline cpp/src/io/parquet/page_enc.cu:1413]: These three can be combined into a single if block to reduce number of __syncthreads.

File: cpp/src/io/orc/stripe_init.cu (hunk 4 above)
Comment (devavret) [reply cpp/src/io/orc/stripe_init.cu:37]: The code below seems to use it interchangeably. Be careful when updating.

File: cpp/src/io/orc/stripe_init.cu (hunk 5 above)
Comment (rgsl888prabhu) [reply cpp/src/io/orc/stripe_init.cu:124]: When you meant bug, do you mean this https://github.com/rapidsai/cudf/pull/6559#discussion_r509831571 ?

Yes, it was.

Comment (rgsl888prabhu) [reply cpp/src/io/orc/stripe_init.cu:124]: And I just removed unnecessary volatile usages in some places, which triggered rearrangement from clang.

Comment (devavret) [reply cpp/src/io/orc/stripe_init.cu:124]: That's parquet. This is ORC. I think they're unrelated. And I think it wouldn't break now. Earlier the cooperative loaders were written thinking they would be read by the entire block but that might not be correct. Not since it's only one thread, instead of one thread of the entire block, this would be read by one thread of each warp.

Comment (rgsl888prabhu) [reply cpp/src/io/orc/stripe_init.cu:124]: Ahh, I missed that. Will take care of it. I was in assumption that `t` was always thread ID, didn't observe it being used as lane id. Will check all other occurrences and correct them.

**Comment by vuule** [comment]
@rgsl888prabhu  Please post the impact of this PR on affected benchmarks.

File: cpp/src/io/orc/stripe_init.cu [hunk 7]
```diff
@@ -32,14 +32,12 @@ extern "C" __global__ void __launch_bounds__(128, 8) gpuParseCompressedStripeDat
 {
   __shared__ compressed_stream_s strm_g[4];
 
-  volatile compressed_stream_s *const s = &strm_g[threadIdx.x >> 5];
-  int strm_id                           = blockIdx.x * 4 + (threadIdx.x >> 5);
-  int t                                 = threadIdx.x & 0x1f;
+  compressed_stream_s *const s = &strm_g[threadIdx.x >> 5];
+  int strm_id                  = blockIdx.x * 4 + (threadIdx.x >> 5);
+  int lane_id                  = threadIdx.x & 0x1f;
```
Comment (jrhemstad) [inline cpp/src/io/orc/stripe_init.cu:37]: This generates the same SASS and is 100x more readable. 
```suggestion
  int lane_id                  = threadIdx.x % 32;
```

File: cpp/src/io/orc/stripe_init.cu [hunk 8]
```diff
@@ -32,14 +32,12 @@ extern "C" __global__ void __launch_bounds__(128, 8) gpuParseCompressedStripeDat
 {
   __shared__ compressed_stream_s strm_g[4];
 
-  volatile compressed_stream_s *const s = &strm_g[threadIdx.x >> 5];
-  int strm_id                           = blockIdx.x * 4 + (threadIdx.x >> 5);
-  int t                                 = threadIdx.x & 0x1f;
+  compressed_stream_s *const s = &strm_g[threadIdx.x >> 5];
```
Comment (jrhemstad) [inline cpp/src/io/orc/stripe_init.cu:35]:
```suggestion
  compressed_stream_s *const s = &strm_g[threadIdx.x / 32 ];
```

File: cpp/src/io/orc/stripe_init.cu [hunk 9]
```diff
@@ -32,14 +32,12 @@ extern "C" __global__ void __launch_bounds__(128, 8) gpuParseCompressedStripeDat
...
+  int strm_id                  = blockIdx.x * 4 + (threadIdx.x >> 5);
```
Comment (jrhemstad) [inline cpp/src/io/orc/stripe_init.cu:36]:
```suggestion
  int strm_id                  = blockIdx.x * 4 + (threadIdx.x / 32);
```

File: cpp/src/io/orc/stripe_init.cu [hunk 10]
```diff
@@ -419,21 +409,13 @@ extern "C" __global__ void __launch_bounds__(128, 8)
...
-      ((volatile uint32_t *)&s->strm_info[t >> 6])[t6] =
-        ((const uint32_t *)&strm_info[strm_id])[t6];
-    }
-  }
   if (t == 0) {
+    s->chunk = chunks[chunk_id];
+    if (strm_info) {
+      s->strm_info[0] = strm_info[s->chunk.strm_id[0]];
+      s->strm_info[1] = strm_info[s->chunk.strm_id[1]];
+    }
```
Comment (davidwendt) [inline cpp/src/io/orc/stripe_init.cu:417]: I'm having trouble seeing where the 
```
s->strm_info[1] = strm_info[s->chunk.strm_id[1]];
```
line comes from in the original code. Was this added to fix a bug?

**Review by davidwendt — CHANGES_REQUESTED** [review]

File: cpp/src/io/orc/stripe_init.cu (hunk 10 above)
Comment (rgsl888prabhu) [reply cpp/src/io/orc/stripe_init.cu:417]: This kernel is of size {128, 1, 1} and `(>>6)` means `/64`,  so earlier two sets of values were being copied. So, I just replaced it with direct update.

Comment (davidwendt) [reply cpp/src/io/orc/stripe_init.cu:417]: `sizeof(CompressedStreamInfo)` looks to be 64 bytes and divided by `sizeof(uint32_t)=4` is 16. 
So doesn't `t6` range from `0-15`?

Comment (rgsl888prabhu) [reply cpp/src/io/orc/stripe_init.cu:417]: But, we are concerned about `s->strm_info[t >> 6]`, which would be only two sets of values in 128 threads

**Review by davidwendt — APPROVED** [review]

File: cpp/src/io/avro/avro_gpu.cu [hunk 11]
```diff
@@ -261,11 +261,7 @@ extern "C" __global__ void __launch_bounds__(NWARPS * 32, 2)
   } else {
```
Comment (vuule) [inline cpp/src/io/avro/avro_gpu.cu:261]: @rgsl888prabhu  Aren't lines 256-257 also cooperative loading?

Comment (rgsl888prabhu) [reply cpp/src/io/avro/avro_gpu.cu:261]: Ahh, missed that. Will push changes.

**Review by devavret — APPROVED** [review]

**Review by vuule — CHANGES_REQUESTED** [review]
requesting changes for a previously made comment.

**Review by vuule — APPROVED** [review]

**Comment by vuule** [comment]
rerun tests

File: cpp/src/io/avro/avro_gpu.cu [hunk 12]
```diff
@@ -250,22 +257,13 @@ extern "C" __global__ void __launch_bounds__(NWARPS * 32, 2)
...
   // Fetch schema into shared mem if possible
   if (schema_len <= MAX_SHARED_SCHEMA_LEN) {
-    for (int i = threadIdx.y * 32 + threadIdx.x;
-         i < schema_len * sizeof(schemadesc_s) / sizeof(uint32_t);
-         i += NWARPS * 32) {
-      reinterpret_cast<uint32_t *>(&g_shared_schema)[i] =
-        reinterpret_cast<const uint32_t *>(schema_g)[i];
-    }
-    __syncthreads();
+    cuda::memcpy_async(g_shared_schema, schema_g, schema_len * sizeof(schemadesc_s), barrier);
```
Comment (jrhemstad) [inline cpp/src/io/avro/avro_gpu.cu:260]: This doesn't look right. The original code copied from `schema_g` cooperatively. This code has every thread redundantly copying the entire `schema_g` array. In order to cooperatively copy you need to pass the `group` as the first argument.

Comment (rgsl888prabhu) [reply cpp/src/io/avro/avro_gpu.cu:260]: Ahh, but passing group is not yet supported as it is depending on #6275 .

Comment (rgsl888prabhu) [reply cpp/src/io/avro/avro_gpu.cu:260]: How about using `memcpy` for now with thread 0 and sync. Once #6275 is merged, we can update this.

Comment (rgsl888prabhu) [reply cpp/src/io/avro/avro_gpu.cu:260]: @jrhemstad Else we can leave it as it before this change and update it once #6275 is merged.

Comment (jrhemstad) [reply cpp/src/io/avro/avro_gpu.cu:260]: You don't want to use normal `memcpy` here. You'd be better off just doing normal element-wise assignment in parallel.
